src/extensions/extension_data/key_share.rs

```diff
@@ -0,0 +1,153 @@
+use heapless::Vec;
+
+use crate::{
+    buffer::CryptoBuffer,
+    extensions::extension_data::supported_groups::NamedGroup,
+};
+
+use crate::{
+    parse_buffer::{
+        ParseBuffer,
+        ParseError,
+    },
+    TlsError,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct KeyShareServerHello<'a>(pub KeyShareEntry<'a>);
+
+impl<'a> KeyShareServerHello<'a> {
+    pub fn parse(buf: &mut ParseBuffer<'a>) -> Result<Self, ParseError> {
+        Ok(KeyShareServerHello(KeyShareEntry::parse(buf)?))
+    }
+
+    pub fn encode(&self, buf: &mut CryptoBuffer) -> Result<(), TlsError> {
+        self.0.encode(buf)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct KeyShareClientHello<'a, const N: usize> {
+    pub client_shares: Vec<KeyShareEntry<'a>, N>,
+}
+
+impl<'a, const N: usize> KeyShareClientHello<'a, N> {
+    pub fn parse(buf: &mut ParseBuffer<'a>) -> Result<Self, ParseError> {
+        //let len = buf.read_u16()? as usize;
+        let len = buf.remaining();
+        Ok(KeyShareClientHello {
+            client_shares: buf.read_list(len, KeyShareEntry::parse)?,
+        })
+    }
+
+    pub fn encode(&self, buf: &mut CryptoBuffer) -> Result<(), TlsError> {
+        // FIXME: RFC states the following, but enconding the length breaks integration tests.
+
+        // In the ClientHello message, the "extension_data" field of this
+        // extension contains a "KeyShareClientHello" value:
+        //
+        // struct {
+        //     KeyShareEntry client_shares<0..2^16-1>;
+        // } KeyShareClientHello;
+
+        //buf.with_u16_length(|buf| {
```

Comment:
I was completely surprised by this while I was working on the changes and I haven't checked since. If anyone happens to know whether I'm just missing some info regarding opaque buffers in opaque buffers, feel free to let me know. Trying to interpret the RFC at 11pm isn't something I expect to do especially well.

Reply:
KeyShareEntry had a length wrapper that was incorrect. Removed that, uncommented this, format should be fine now.